### ISSUE TITLE
Input weights from last training as initial weights

### ISSUE THREAD

**Issue by jyotikr** [yhenon/keras-frcnn#187]
Hello I have a question. Is it possible to input weights from last training as initial weights in order to improve the model?

I do see that as an option while training the model so I assume answer is Yes. So my other question is how do i know that the initial weights has been read correctly? I thought if I have given good initial weight as starting point from previous training than my Total loss should have been finite number instead of Inf which I have noticed.

Let me know if my question is not clear.

**Comment by Tbarowski** [comment]
In general: Resuming training is possible. Please consider that more training does not mean better results. There is an optimum before overfitting which should be determined by validation loss (which is not implemented in the current version). And yes, for best performance you should start with the last loss.

An easy possibility to debug the loading process is the following: load the weights, save the model to h5 directly and compare the weights of some layers using a hdf5 viewer (like hdfview).

**Comment by jyotikr** [comment]
Thank you Tbarowski for your response. Yes I understand more training does not always mean better result.

I have another clarification. What happens if I resume the training after adding few more training images. I am asking this because i observed some wrong detection on my test dataset. I thought i can re-label them and include these in my training dataset. But ofcourse if I can avoid re-training and start from where i stopped last it will be great.

**Comment by Tbarowski** [comment]
Adding examples in general is no problem from a training performance view.  Adding examples from your testing set is a different thing. Your test scores will increase but your test dataset loses kind of its intended function. If you view it from the application side it will of course perform better, if you feed him this hard examples.

In my opinion you should add images that share the same characteristics as the ones that have bad performance but are not exactly the same. Then your test set is independent. But it also depends on your task you want to solve (very specific <-> very general).

**Comment by shinstra** [comment]
Another thing to be aware of if you are doing start-stop training like you described is that you will also need to save the optimizer state as well as the weights - can be done by changing `model_all.save_weights(C.model_path)` to `model_all.save_model(C.model_path)`. Otherwise your learning rate will likely be initially too high in successive training runs.

**Comment by Tbarowski** [comment]
>  can be done by changing model_all.save_weights(C.model_path) to model_all.save_model(C.model_path).

it's just model_all.save(C.model_path)

**Comment by shinstra** [comment]
Ah, right you are.

**Comment by jyotikr** [comment]
Thank you both, Tbarowski and shinstra. This was really useful.

**Comment by Tbarowski** [comment]
If you want to load the rpn and the classifier model for training you have to consider two things:

First, see my Pull Request ([#link](https://github.com/yhenon/keras-frcnn/pull/194)) on RoiPoolingConv.py which is required to load the complete model.

Second, when loading the model you have to pass the custom layers and losses like
model = load_model('my_model.h5', custom_objects={'RoiPoolingConv': RoiPoolingConv})


model_rpn2 = load_model('/yourpath/model_rpn.h5', custom_objects={'RoiPoolingConv.py': RoiPoolingConv,
                                                                          'FixedBatchNormalization': FixedBatchNormalization,
                                                                          'rpn_loss_regr_fixed_num': rpn_loss_regr(num_anchors),
                                                                          'rpn_loss_cls_fixed_num': rpn_loss_cls(num_anchors)})

**Comment by yhenon** [comment]
@Tbarowski I've merged that PR btw

**Comment by Tbarowski** [comment]
Yhenon, do you have any experience with loading the whole model instead just the weights? I'm not sure if the shared weights will work the same, if we load all three models with load_model().

Any ideas on that?

**Comment by giladdiv** [comment]
@Tbarowski,Hi, have you encounter a problem with loading the BN parameters? I am trying to save weights from one model and load it the other model with some trainable = False layers,and all the weights are loaded fine except the BN parameters.
